dbms/src/Storages/DeltaMerge/Decode/SSTFilesToBlockInputStream.cpp

```diff
@@ -139,23 +141,29 @@ SSTFilesToBlockInputStream::~SSTFilesToBlockInputStream() = default;
 
 void SSTFilesToBlockInputStream::readPrefix() {}
 
-void SSTFilesToBlockInputStream::checkFinishedState(SSTReaderPtr & reader)
+void SSTFilesToBlockInputStream::checkFinishedState(SSTReaderPtr & reader, ColumnFamilyType cf)
 {
     // There must be no data left when we write suffix
     if (!reader)
         return;
     if (!reader->remained())
         return;
-    RUNTIME_CHECK_MSG(soft_limit.has_value(), "soft_limit.has_value()");
+    if (prehandle_task->abort_flag.load())
+        return;
+
+    RUNTIME_CHECK_MSG(soft_limit.has_value(), "soft_limit.has_value(), cf={}", magic_enum::enum_name(cf));
     BaseBuffView cur = reader->keyView();
-    RUNTIME_CHECK_MSG(buffToStrView(cur) > soft_limit.value().raw_end, "cur > raw_end");
+    RUNTIME_CHECK_MSG(
+        buffToStrView(cur) > soft_limit.value().raw_end,
+        "cur > raw_end, cf={}",
+        magic_enum::enum_name(cf));
```

Comment:
As we will call `reader->seekToLast();` in `maybeStopBySoftLimit`. Is this check still necessary?
https://github.com/pingcap/tiflash/blob/33ac2fd7e7c253882f9650499ababdf769485bd8/dbms/src/Storages/DeltaMerge/Decode/SSTFilesToBlockInputStream.cpp#L516-L517

Reply:
This check guards the invariance property that all data is properly handled when `readSuffix`. If this check fires, there must be some newly added/modified logic that changed this behavior and fails to handle all the data, which can lead to data inconsistence. The logic of consuming this stream depends on this invariance property.

`reader->seekToLast()` in `maybeStopBySoftLimit` means all the rest data is no longer necessary, so we can stop read and seek to last, which is a behavior this check forces. If we don't conform to this behavior, then there can be some abundant calls, if someone only checks `->remained()`.

Why not just put the `soft_limit` into `SSTReader`, and make the `->remained` the only check for ending? We don't do this since `soft_limit` is a hint that co-works with the encoding of TiKVKey, however, the SSTReader only handles plain string. So IMO the best place for `soft_limit` is `SSTFilesToBlockInputStream`.